plugins/@grouparoo/zendesk/src/lib/export/exportProfile.ts

```diff
@@ -0,0 +1,253 @@
+      updated = await client.users.createOrUpdate({ user: payload });
+    }
+
+    if (email && updated.email !== email) {
+      // console.log("updating email", `${updated.email} -> ${email}`);
+      // have to make this the primary
+      // https://developer.zendesk.com/rest_api/docs/support/users#email-address
+      await makeEmailPrimary(client, updated.id, email);
+    }
+
+    return { success: true };
+  } catch (error) {
+    // https://develop.zendesk.com/hc/en-us/articles/360001074328-Best-practices-for-avoiding-rate-limiting
+
+    // headers
+    // X-Rate-Limit: 700
+    // X-Rate-Limit-Remaining: 699
+
+    // The response contains the following key information:
+    // error.statusCode = 429;
+    // error.retryAfter = retryAfter;
+    if (error?.statusCode === 429) {
+      let retryIn = error.retryAfter; // seconds
+      if (!retryIn || retryIn < 0) {
+        retryIn = 60; // per minute
+      }
+      // add some random time to that to spread it out
+      retryIn += Math.floor(Math.random() * 30) + 1;
+      return { error, success: false, retryDelay: 1000 * retryIn };
+    }
+
+    throw error;
+  }
+};
+
+export async function searchForUser(client, findBy: any) {
+  const response = await client.users.search(findBy);
+  let found = null;
+  for (const user of response) {
+    if (user.active) {
+      if (!found) {
+        // not deleted
+        found = user;
+      } else {
+        throw `more than one id result! ${JSON.stringify(findBy)}: ${
+          user.id
+        } and ${found.id}`;
+      }
+    }
+  }
+  return found;
+}
+export async function findUser(
+  client,
+  newProfileProperties,
+  oldProfileProperties
+) {
+  const newId = newProfileProperties.external_id;
+  const oldId = oldProfileProperties.external_id;
+
+  let found = null;
+  if (!found && newId) {
+    found = await searchForUser(client, { external_id: newId });
+  }
+  if (!found && oldId && oldId !== newId) {
+    found = await searchForUser(client, { external_id: oldId });
+  }
+  return found;
+}
+
+async function getRootFields(client) {
+  const required = getRequiredFields();
+  const builtIn = getBuiltInFields();
+
+  const out = {};
+  for (const field of required) {
+    out[field.key] = true;
+  }
+  for (const field of builtIn) {
+    out[field.key] = true;
+  }
+  return out;
+}
+
+function formatVar(value) {
+  if (!value) {
+    return null;
+  }
+  if (value instanceof Date) {
+    return value.toISOString();
+  } else {
+    return value;
+  }
+}
+
+function getDefaultName(payload) {
+  const priority = ["alias", "email", "external_id"];
+  for (const key of priority) {
+    const value = (payload[key] || "").toString().trim();
+    if (value.length > 0) {
+      return value;
+    }
+  }
+  return "Unknown";
+}
+
+function makeTagName(groupName) {
+  // tags can't have spaces and have to be lowercase
+  let tagName = groupName.toLowerCase();
+  tagName = tagName.replace(/[^a-z]/g, "_");
+  return tagName;
+}
```

Comment:
this is starting to appear rather common - we could enforce this in our group names...

Reply:
I made a task to DRY up code that's common. and have this in it.
T-372